lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,10 +12,92 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
+  ExprMap m_cache;
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  }
+
+  bool isVar(Expr exp) { return bind::isBoolVar(exp) || bind::isIntVar(exp); }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
+
+  Expr inferTypeITE(Expr exp) {
+    Expr boolSort = sort::boolTy(exp->efac());
+
+    // ite(a,b,c) : a is bool type, b and c are the same type
+    if (exp->arity() == 3 && m_cache.at(exp->arg(0)) == boolSort &&
+        (m_cache.at(exp->arg(1)) == m_cache.at(exp->arg(2))))
+      return boolSort;
+    else
+      return Expr();
+  }
+
+  // ensures the expression has the correct number of children and all children
+  // are bool types
+  Expr boolCheckChildren(Expr exp, bool (*checkNumChildren)(int)) {
+    auto isBool = [this](Expr exp) {
+      return this->m_cache.at(exp) == sort::boolTy(exp->efac());
```

Comment:
don't create so many expressions. use `isOp<BOOL_TY>()` instead. Cache types in type checker if needed.